src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs

```diff
@@ -0,0 +1,267 @@
+            return stream;
+        }
+
+        private async ValueTask<Stream?> GetEmbeddedFileAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            string packagePath = uri.LocalPath;
+            if (File.Exists(packagePath))
+            {
+                string fileRelativePath = PathUtility.StripLeadingDirectorySeparators(
+                Uri.UnescapeDataString(uri.Fragment)
+                    .Substring(1)); // Substring skips the '#' in the URI fragment
+
+                string dirPath = Path.GetDirectoryName(packagePath);
+                string extractedIconPath = Path.Combine(dirPath, fileRelativePath);
+
+                // use GetFullPath to normalize "..", so that zip slip attack cannot allow a user to walk up the file directory
+                if (Path.GetFullPath(extractedIconPath).StartsWith(dirPath, StringComparison.OrdinalIgnoreCase) && File.Exists(extractedIconPath))
+                {
+                    Stream fileStream = new FileStream(extractedIconPath, FileMode.Open);
+                    return fileStream;
+                }
+                else
+                {
+                    try
+                    {
+                        using (PackageArchiveReader reader = new PackageArchiveReader(packagePath))
+                        using (Stream parStream = await reader.GetStreamAsync(fileRelativePath, cancellationToken))
+                        {
+                            var memoryStream = new MemoryStream();
+                            await parStream.CopyToAsync(memoryStream);
+                            memoryStream.Seek(0, SeekOrigin.Begin);
+                            return memoryStream;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Debug.WriteLine(ex);
+                        return null;
+                    }
+                }
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private async Task<Stream?> GetStream(Uri uri)
+        {
+            // BitmapImage can download on its own from URIs, but in order
+            // to support downloading on a worker thread, we need to download the image
+            // data and put into a memorystream. Then have the BitmapImage decode the
+            // image from the memorystream.
+
+            byte[]? imageData = null;
+            MemoryStream? memoryStream = null;
+
+            if (uri.IsFile)
+            {
+                if (File.Exists(uri.LocalPath))
+                {
+                    memoryStream = new MemoryStream(File.ReadAllBytes(uri.LocalPath));
+                    return memoryStream;
+                }
+                else
+                {
+                    return null;
+                }
+            }
+            else
+            {
+                using (var httpClient = new HttpClient())
```

Comment:
Don't we need to create just one instance of HttpClient and reuse it? In general HttpClient dispose doesn't really dispose and some point cause memory leak.  We can lazy init instance and keep using since it might get called many times.

https://stackoverflow.com/a/44959403  Only problem with this approach is DNS change mentioned here.

Reply:
i'm replacing webclient usage with httpclient for icon loading. 
webclient - we used an instance per fetch.
httpclient - i used the same approach.

do we share one httpclient for our restore codepath too?

happy to move towards 1 instance if that makes the most sense.

Reply:
@rrelyea 
Not sure about restore codepath. But for now we don't have to share httpClient with restore codepath.  This one can have just 1 dedicated http instance.
In my previous project this HttpClient thing caused memory leak one moment we had hundreds of them , that is why I'm concerned.

Reply:
copying approach from httpsource.cs for this. makes sense to do. thx.

Reply:
actually, not sure why i need to delay creation of httpclient, so not using same approach as httpsource.

Reply:
Sure, httpsource implementation lgfm.

Reply:
did a simple field set that happens at construction to the new httpclient.
could do something like httpsource.cs (locks, delay create) if people see wisdom to do that in this case. 
(not sure what lgfm is... did you mean lgtm?)

Reply:
Yes, looks good for me.